### ISSUE TITLE
NNBD: Dynamic soundness checking due to opted-in/out code is unspecified

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#624]
Consider the following situation:

```dart
// Opted-in code in 'opt-in.dart'.
class A { String get s => "Hello!"; }

// Opted-out code in 'opt-out.dart'.
import 'opt-in.dart';

class B extends A { String get s => null; }

// Opted-in main library.
import 'opt-in.dart';
import 'opt-out.dart';

main() {
  A a = B();
  print(a.s.length);
}
```

We have the [following rule](https://github.com/dart-lang/language/blob/master/accepted/future-releases/nnbd/feature-specification.md#errors-as-warnings) in the nnbd specification:

> When strong null checking is enabled, errors specified in this
> proposal shall be treated as errors

But opted-out libraries [must be checked](https://github.com/dart-lang/language/blob/master/accepted/future-releases/nnbd/feature-specification.md#legacy-libraries) according to an erased version of any opted-in libraries that they import:

>  All opted-in libraries upstream from the legacy library are viewed
> by the legacy library with nullability related features erased from their APIs.

This means that it is not an error for class `B` to have the declaration of `s` as in the example; it is also not an error for class `A` to be declared as it is; and it is not an error to have the invocation in `main` of `a.s.length`.

So the point is (as we know) that NNBD-and-legacy-code-together is not sound, and it is not specified how eagerly we will detect an unsoundness and raise a dynamic error (if at all).

In particular, would we expect a caller-side check and a null failure in the following variant?:

```dart
main() {
  A a = B();
  dynamic d = a.s;
}
```

With a static type of `String`, it is a soundness violation for `a.s` to evaluate to the null object. So it would be an acceptable approach to handle the null-soundness violation by throwing when the value of `a.s` is obtained, even though the initialization of `d` will not fail.

However, we can only determine that there is a need for a dynamic check on the value of `a.s` if we perform a global analysis and find the declaration of `B`. So we might very well want to avoid performing this kind of caller-side check (for performance reasons). Otherwise we would basically need a dynamic check on every expression (including every subexpression of any given expression), if its static type is potentially non-nullable.

The same problem arises at a higher level with type arguments: If a variable of type `List<int>` due to opt-in/out unsoundness at some point refers an instance of `List<int?>` (where we may not even have any elements whose value is null), is the runtime required to perform a check such that there is a dynamic error, or would it be OK for the program to use that list and never see a problem, because it contains only non-null elements, and no other operations give rise to a direct type check that this object is actually a `List<int>`?

```dart
// Opted-in code in 'opt-in-2.dart'.
List<int> xs;
List<int?> ys = [1, 2, 3];

// Opted-out code in 'opt-out-2.dart'.
import 'opt-in-2.dart';
void foo() { xs = ys; }

// Opted-in main.
import 'opt-in-2.dart';
import 'opt-out-2.dart';

main() {
  foo();
  int sum = 0;
  for (int i in xs) { sum += i; }
}
```

It seems reasonable to have a lazy approach to checks for this kind of null-soundness violations.

However, if we do this then it will be rather ambiguous to test whether a given implementation has a correct behavior, so presumably every test of this kind of situation would need to contain several operations such that the laziness is guaranteed to have been exhausted. For instance, testing that `a.s` throws in the above example is not enough, but testing that `a.s.length` throws is OK because no amount of laziness will allow that to succeed.

**Comment by leafpetersen** [comment]
I'm not sure what this issue is proposing.  The NNBD specification requires no implicit runtime checking for nulls, except for the proposal to insert assertions in debug mode in certain situations.  The semantics of the initial example are to throw a `NoSuchMethodError` on the invocation of `.length` on `null`.

**Comment by eernstg** [comment]
@leafpetersen wrote:
> I'm not sure what this issue is proposing.

I just noticed that we have caller-side checks for certain unsafe usages (like a getter returning `void Function(X)` in a class `C<X>`). So does this mean that we should have caller-side checks for a similar typing situation involving nullability?

And I suggested that we should not:

> It seems reasonable to have a lazy approach to checks for this
> kind of null-soundness violations.

(lazy means that we don't perform the caller-side check, and the trade-off is that we might get an error later because we try to use a member of some non-`Null` class, and the receiver turns out to be null).

I think we agree that we shouldn't have these caller-side nullability checks.
